Fix translate doesn't work on CHAR values

albericgenius opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Fix #14038
Fix translate doesn't work on CHAR values

This PR is used to proposal the solution, there are some test cases not passed.
if the direction is correct, I will fix the test cases asap.

<!-- Provide a user-friendly explanation, keep it brief if it isn't user-visible. -->
## Non-technical explanation



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`14038`)
```

albericgenius commented on core/trino-parser/src/main/java/io/trino/sql/tree/CharLiteral.java:
```diff
@@ -41,7 +39,7 @@ public CharLiteral(Optional<NodeLocation> location, String value)
         super(location);
         requireNonNull(value, "value is null");
         this.length = value.length();
-        this.value = CharMatcher.is(' ').trimTrailingFrom(value);
```
@findepi 
want to learn from you, i do not know why we need to trimRight for char?
so
```
SELECT translate(CHAR 'A BC ', ' ', '_');
we transfer to in prepareQuery
SELECT translate(CHAR 'A BC', ' ', '_');
```
so could we remove this.

findepi replied:
The char type knows its (exactl) length, so the _internal_ representation of char values has trailing spaces omitted to save space.
From end-user perspective this space-related optimization is invisible.

albericgenius replied:
@findepi Thanks for sharing this, I went a wrong direction at the beginning. But finally I got the right solution, I think. :). And i learned a lot during this.
Thanks

albericgenius commented:
@findepi if we need to replace the deprecated assertFunction by ```assertThat(assertions.function``` in TestStringFunctions.java. 
If there is open issue for that, could you help to assign to me.
Thanks

albericgenius commented:
commit to fix the CI issue.

albericgenius commented:
> @albericgenius, there's no issue for it, but feel free to do it if you want and have time. There are dozens of classes that still rely on the deprecated assertFunction and any help to migrate those is appreciated. Take a look at #14528 for some examples.

@martint I have a look at #14528, just do not know where I could start, i am interested in working on this. please feel free to assign the task to me.

@findepi I have rebased the lasted code, but there is still one Product test failed because of "Test execution exceeded timeout of 2.00h". Which is not relative to my commit. I just added a translate function for char.

I have change to use assertFunction in the test, when we migrate to new assertions, i will do it later.

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/StringFunctions.java:
```diff
@@ -915,6 +914,21 @@ public static boolean startsWith(@SqlType("varchar(x)") Slice source, @SqlType("
+    @ScalarFunction("translate")
+    @LiteralParameters({"x", "y", "z"})
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice charTranslate(@LiteralParameter("x") Long x,
```
`translateChar`

findepi commented on core/trino-main/src/test/java/io/trino/operator/scalar/TestStringFunctions.java:
```diff
@@ -968,6 +968,8 @@ public void testTranslate()
         assertFunction("translate('abcd', 'a', 'z')", VARCHAR, "zbcd");
         assertFunction("translate('abcda', 'a', 'z')", VARCHAR, "zbcdz");
 
+        assertFunction("translate(CHAR 'a bc ', ' ', '_')", VARCHAR, "a_bc_");
```
add cases with trailing spaces in the result

```
        assertFunction("translate(CHAR 'abc', 'c', ' ')", VARCHAR, "ab ");
        assertFunction("translate(CHAR 'abc ', 'c', 'z')", VARCHAR, "abz ");
```

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/StringFunctions.java:
```diff
@@ -899,7 +899,6 @@ public static Slice concat(@LiteralParameter("x") Long x, @SqlType("char(x)") Sl
         Slice result = Slices.allocate(leftLength + rightLength);
         result.setBytes(0, paddedLeft);
         result.setBytes(leftLength, right);
-
```
please avoid unrelated changes

findepi commented:
Code looks generally OK, but per https://github.com/trinodb/trino/issues/14038#issuecomment-1239531165 i think the proper fix would be to do https://github.com/trinodb/trino/issues/9031.

albericgenius commented:
> > I have tried to fix via char/varchar coercion in the beginning. But seems it affected a lot of functions,
> 
> yes, this is not easy

Hi @findepi 
I try to work on #9031 as following. I have went though EXACT and IMPLICIT_COERCION resolveFunction logic.
```
+ return Optional.of(createVarcharType(((CharType) sourceType).getLength())); in coerceTypeBase function
+ use the min length for bindings. 
```
But for typeCoercion.getCommonSuperType, I do not know how to fit all requirement(all existed functions). So i probably track on #9031 and learn how to address this issue.

For current commit, add translateChar, for my side, it is good to merge, because for the case of `translate(CHAR 'a bc ', CHAR ' ', CHAR '_')`, translateChar will be hit by EXACT resolveFunction first, it will be more efficient compare to IMPLICIT_COERCION. What is your opinion?

albericgenius commented:
@martint sorry to bother again, I think this fix is good for `translate(CHAR 'a bc ', CHAR ' ', CHAR '_')` case, so still need your help to review this.

martint reviewed (commented):
This is an ok workaround until the underlying issue is fixed.

albericgenius commented on core/trino-main/src/main/java/io/trino/operator/scalar/StringFunctions.java:
```diff
@@ -915,6 +915,23 @@ public static boolean startsWith(@SqlType("varchar(x)") Slice source, @SqlType("
+    @ScalarFunction("translate")
+    @LiteralParameters({"x", "y", "z"})
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice translateChar(@LiteralParameter("x") Long x,
```
Thanks for your help. :)

albericgenius commented on core/trino-main/src/main/java/io/trino/operator/scalar/StringFunctions.java:
```diff
@@ -915,6 +915,23 @@ public static boolean startsWith(@SqlType("varchar(x)") Slice source, @SqlType("
+    @Description("Translate characters from the source string based on original and translations strings")
+    @ScalarFunction("translate")
+    @LiteralParameters({"x", "y", "z"})
+    @SqlType(StandardTypes.VARCHAR)
```
translate(VARCHAR 'sdf' ....  will go translate(varchar(x)....
translate(CHAR 'sdf' ... will go translateChar(char(x) ....
:) if I change char(x) to varchar(x), we can not resolveFunction for (translate(char(x),char(y),char(z))

albericgenius commented:
@martint :) , I addressed the code style issues, Thanks for share this for me, if there any other issue, I will update as soon as possible.

martint commented:
Sorry for the late reply. Can you please add a few tests with various combinations of CHAR and VARCHAR? The existing tests only cover all values being of the same type or only one value being CHAR.

For instance, we should test `translate(CHAR, CHAR, VARCHAR)`, `translate(CHAR, VARCHAR, VARCHAR)`, `translate(VARCHAR, CHAR, VARCHAR)`, and so on.

zhaner08 commented:
> Sorry for the late reply. Can you please add a few tests with various combinations of CHAR and VARCHAR? The existing tests only cover all values being of the same type or only one value being CHAR.
> 
> For instance, we should test `translate(CHAR, CHAR, VARCHAR)`, `translate(CHAR, VARCHAR, VARCHAR)`, `translate(VARCHAR, CHAR, VARCHAR)`, and so on.

Just curious, by testing do you mean Trino wants to support those permutations as well or those tests should be added to ensure other permutations are not supported??

albericgenius commented:
@findepi @martint  you are right, when char 'z', the z.intValue() is 1, when varchar 'z', the z.intValue() is 65536. it have not resolved the root cause.

mosabua commented:
:wave: @albericgenius @findepi @martint  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
